File: 426_convert_binary_search_tree_to_sorted_doubly_linked_list/Solution.java
class Node {
    public int val;
    public Node left;
    public Node right;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val,Node _left,Node _right) {
        val = _val;
        left = _left;
        right = _right;
    }
}

class Solution {
    public static Node createTreeUsingBFS(int[] nums) {
        Node[] nodeArr = new Node[nums.length];
        for (int i = 0; i < nums.length; i++) {
            nodeArr[i] = new Node(nums[i]);
            if (i == 0) continue;
            if (i % 2 == 1) nodeArr[(i - 1) / 2].left = nodeArr[i];
            else nodeArr[(i - 1) / 2].right = nodeArr[i];
        }

        return nodeArr[0];
    }

    public static void printTree(Node root) {
        if (root == null) return;
        printTree(root.left);
        System.out.print(root.val + " ");
        printTree(root.right);
    }

    public static void printList(Node head) {
        Node tmp = head;
        do {
            System.out.print(tmp.val + " ");
            tmp = tmp.right;
        } while (tmp != head);
        System.out.println();
    }

    Node pre, head;
    public Node treeToDoublyList(Node root) {
        if (root == null) return null;
        dfs(root);
        head.left = pre;
        pre.right = head;

        return head;
    }

    private void dfs(Node cur) {
        if (cur == null) return;
        dfs(cur.left);
        if (pre != null) pre.right = cur;
        else head = cur;
        cur.left = pre;
        pre = cur;
        dfs(cur.right);
    }

    public static void main(String[] args) {
        Node root = createTreeUsingBFS(new int[]{10, 5, 15, 2, 6, 11, 16, 1, 3});
        printTree(root);
        System.out.println();
        Node head = new Solution().treeToDoublyList(root);
        printList(head);
    }
}

File: 426_convert_binary_search_tree_to_sorted_doubly_linked_list/Solution.go
package main

import "fmt"

type Node struct {
    Val   int
    Left  *Node
    Right *Node
}

func createTreeUsingBFS(nums []int) *Node {
    var nodeArr []*Node

    for i, num := range nums {
        nodeArr = append(nodeArr, &Node{num, nil, nil})
        if i == 0 { continue }
        if i % 2 == 1 {
            nodeArr[(i - 1) / 2].Left = nodeArr[i]
        } else {
            nodeArr[(i - 1) / 2].Right = nodeArr[i]
        }
    }

    return nodeArr[0]
}

func printTree(root *Node) {
    if root == nil { return }
    printTree(root.Left)
    fmt.Printf("%d ", root.Val)
    printTree(root.Right)
}

func printList(head *Node) {
    tmp := head
    for {
        fmt.Printf("%d ", tmp.Val)
        tmp = tmp.Right
        if tmp == head { break }
    }
    fmt.Println()
}

var pre, head *Node
func treeToDoublyList(root *Node) *Node {
    if root == nil { return nil }

    dfs(root)
    head.Left = pre
    pre.Right = head

    return head
}

func dfs(cur *Node) {
    if cur == nil { return }
    dfs(cur.Left)
    if pre != nil {
        pre.Right = cur
    } else {
        head = cur
    }
    cur.Left = pre
    pre = cur
    dfs(cur.Right)
}

func main() {
    root := createTreeUsingBFS([]int{10, 5, 15, 2, 6, 11, 16, 1, 3})
    printTree(root)
    fmt.Println()
    head := treeToDoublyList(root)
    printList(head)
}
